### PR TITLE
Move to using C99 int types and deprecating zephyr int types

### LINKED ISSUE #25327: Move to C99 integer types and deprecate zephyr specific types
With the introduction of the following commit:

commit f32330b22c333012cf5f771b0eae9ef378e446d4
Author: Nicolas Pitre <npitre@baylibre.com>
Date:   Wed Jun 5 12:19:37 2019 -0400

    stdint.h: streamline type definitions

We've made the C99 and Zephyr types (`include/zephyr/types.h`) match definition (for the most part [caveats below]).   We've been intermixing C99 and the Zephyr types in the Zephyr code base for a little while now.  Since Zephyr ends up link with other projects (mbedtls, HALs, lvgl, mcuboot, etc..) utilizing the C99 types will be a bit more consistent going forward.  We introduced the zephyr types to try and keep sized integer types consistent across all architectures.  The intent of this change is to try and maintain that goal as much as possible.

Caveats:
* `(u)int64_t` wasn't defined to match the definition of `(u)64_t`.  Its easy enough to handle the 64-bit integer cases in `include/toolchain/zephyr_stdint.h` the same way that the 32-bit integers are today.
* `(u)int64_t` will be inconsistent on native_posix_64.  For platforms in which we target real hardware Zephyr controls how the C99 types will be defined (via `include/toolchain/zephyr_stdint.h`).  However the native_posix builds we don't have this ability as we are utilize the host systems libc environment.  It happens that the native_posix (32-bit) build types for the C99 types aligned with how we'll define them in Zephyr.  However on the native_posix_64 target we will be inconsistent for `(u)int64_t` as it will be defined as a `(unsigned) long` [for native_posix_64] and on all other builds it will be `(unsigned) long long`.  We can deal with the few cases that show up w/use of `PRI.64` macros.

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

**Comment by ceolin** [comment]
I'm totally in favor of using standard stdint types .

**Comment by sjg20** [comment]
These are longer and harder to type. Could we do what linux does instead?

u64, u32, u16, u8 for unsigned

**Comment by galak** [comment]
> These are longer and harder to type. Could we do what linux does instead?
> 
> u64, u32, u16, u8 for unsigned

The issue is that a lot of code integrated w/zephyr utilizes C99 and those types have leaked into Zephyr.  If we plan on getting 2 different types, we'd just leave the zephyr types as is.

**Comment by galak** [comment]
The sanitycheck failure is timeout of a test that passes when run by itself.

**Review by marc-hb — COMMENTED** [review]
> Move to using PRIu64/PRId64 instead of %llu/%lld since on native_posix_64 the uint64_t/int64_t type is defined in terms of 'long int' and not 'long long int'.

Dropping the keyword "LP64"  in the commit message and documentation would add a lot of relevant background information with very few extra characters. Something like: "Unlike Zephyr, native_posix_64 follows the LP64 model that blabla..."

https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models
https://www.google.com/search?q=__LP64__

> These are longer and harder to type.

https://twitter.com/study_web_dev/status/764811712394620928

(Not counting all the time spent trying to understand Zephyr types - and I barely did)

> Could we do what linux does instead?

The Linux kernel doesn't look like it will ever care about POSIX-looking code.

File: Kconfig.zephyr [hunk 1]
```diff
@@ -363,6 +363,12 @@ config LEGACY_DEVICETREE_MACROS
 	  Zephyr 2.2 and previous versions, rather than the devicetree.h
 	  API introduced during the Zephyr 2.3 development cycle.
 
+config DEPRECATED_ZEPHYR_INT_TYPES
```
Comment (carlescufi) [inline Kconfig.zephyr:366]: I think we should use `LEGACY`, like we've done for DT and timeouts

File: Kconfig.zephyr [hunk 2]
```diff
@@ -363,6 +363,12 @@ config LEGACY_DEVICETREE_MACROS
...
+	bool "Allow use deprecated zephyr integer types"
```
Comment (carlescufi) [inline Kconfig.zephyr:367]:
```suggestion
	bool "Allow the use of the legacy zephyr integer types"
```

File: Kconfig.zephyr [hunk 3]
```diff
@@ -363,6 +363,12 @@ config LEGACY_DEVICETREE_MACROS
...
+	help
+	  Allows use of the deprecated Zephyr integer typedefs defined in
```
Comment (carlescufi) [inline Kconfig.zephyr:369]:
```suggestion
	  Allows the use of the legacy Zephyr integer typedefs defined in
```

File: Kconfig.zephyr [hunk 4]
```diff
@@ -363,6 +363,12 @@ config LEGACY_DEVICETREE_MACROS
...
+	  Zephyr 2.3 and previous versions.
+
```
Comment (carlescufi) [inline Kconfig.zephyr:371]: Those are `u<n>_t` and `s<n>_t`

File: Kconfig.zephyr (hunk 1 above)
Comment (carlescufi) [reply Kconfig.zephyr:366]:
```suggestion
config LEGACY_ZEPHYR_INT_TYPES
```

**Review by carlescufi — CHANGES_REQUESTED** [review]
Only reviewed the new Kconfig option for now

File: Kconfig.zephyr (hunk 4 above)
Comment (carlescufi) [reply Kconfig.zephyr:371]: @galak I was expecting to mention the types by name as I wrote above.

**Review by nashif — APPROVED** [review]

File: scripts/coccinelle/unsigned_suffix.cocci [hunk 5]
```diff
@@ -9,8 +9,8 @@ virtual patch
 virtual report
 
 @r_unsigned@
-typedef uint8_t, uint16_t, uint32_t, uint64_t, u8_t, u16_t, u32_t, u64_t;
-{unsigned char, unsigned short, unsigned int, uint8_t, uint16_t, uint32_t, uint64_t, u8_t, u16_t, u32_t, u64_t} v;
+typedef uint8_t, uint16_t, uint32_t, uint64_t, uint8_t, uint16_t, uint32_t, uint64_t;
+{unsigned char, unsigned short, unsigned int, uint8_t, uint16_t, uint32_t, uint64_t, uint8_t, uint16_t, uint32_t, uint64_t} v;
```
Comment (stephanosio) [inline scripts/coccinelle/unsigned_suffix.cocci:13]: spurious change

File: scripts/coccinelle/unsigned_suffix.cocci [hunk 6]
```diff
@@ -50,7 +50,7 @@ if C.isdigit() != True:
 coccinelle.z = C + "U"
 
 @r_subst depends on patch@
-{unsigned char, unsigned short, unsigned int, uint8_t, uint16_t, uint32_t, uint64_t, u8_t, u16_t, u32_t, u64_t} r_unsigned.v;
+{unsigned char, unsigned short, unsigned int, uint8_t, uint16_t, uint32_t, uint64_t, uint8_t, uint16_t, uint32_t, uint64_t} r_unsigned.v;
```
Comment (stephanosio) [inline scripts/coccinelle/unsigned_suffix.cocci:53]: spurious change

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: scripts/coccinelle/unsigned_suffix.cocci (hunk 6 above)
Comment (galak) [reply scripts/coccinelle/unsigned_suffix.cocci:53]: fixed

File: scripts/coccinelle/unsigned_suffix.cocci (hunk 5 above)
Comment (galak) [reply scripts/coccinelle/unsigned_suffix.cocci:13]: fixed

File: scripts/coccinelle/unsigned_suffix.cocci (hunk 6 above)
Comment (galak) [reply scripts/coccinelle/unsigned_suffix.cocci]: fixed

**Review by stephanosio — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by Laczen — APPROVED** [review]

**Review by chrta — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by andyross — APPROVED** [review]
I can't believe we're doing this again.  But OK, quick spot checks show that the scripting did the right thing.  Please rip the bandaid off and merge this quickly so we can get past the resulting rebase hell.

File: Kconfig.zephyr (hunk 1 above)
Comment (nashif) [reply Kconfig.zephyr]: Legacy means that you want to maintain this beyond the deprecation time, deprecated is more appropriate IMO, meaning at some point it is going away.

**Comment by nashif** [comment]
I am not fond of calling old types legacy, legacy means that you want to maintain those for a lengthy period of time where deprecated is more strict and follows deprecation policies.

File: Kconfig.zephyr (hunk 4 above)
Comment (carlescufi) [reply Kconfig.zephyr:371]: @galak could you please mention the types by name here?

**Review by carlescufi — CHANGES_REQUESTED** [review]
Please add an entry in the API Changes section of the 2.4 release notes in this PR

**Comment by carlescufi** [comment]
> I am not fond of calling old types legacy, legacy means that you want to maintain those for a lengthy period of time where deprecated is more strict and follows deprecation policies.

It's what we do when we deprecate via Kconfig option. At least it's what we've done so far (timeouts and DT, which will both go away in 2 releases), so for consistency I think we should keep the same naming scheme.

It's mentioned here: https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#deprecated

**Comment by nashif** [comment]
> t's what we do when we deprecate via Kconfig option. At least it's what we've done so far (timeouts and DT, which will both go away in 2 releases), so for consistency I think we should keep the same naming scheme.
> 
> It's mentioned here: https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#deprecated

well, just because we have what I think is a mistake, does not mean we have to continue with that mistake, and all of this happened just recently...


https://www.linkedin.com/pulse/deprecated-vs-legacy-ayoub-moustaid/

the way I see it:
- deprecated: is something that you should not use and will be removed at some point following the deprecation policy. (2 releases)
- legacy:  won't be removed for backwards compatibility or other reasons.

Are we going to keep those types more than 2 releases?

**Comment by carlescufi** [comment]
> Are we going to keep those types more than 2 releases?

No, I don't think so. So in that case I suggest we change the doc, because right now (mostly because of the timeout and DT Kconfig options using `LEGACY` even though they'll be removed after 2 releases) it recommends using `LEGACY` and not `DEPRECATED` in the Kconfig option name.
I'm not against changing policy, I just want us to be consistent.

**Review by jukkar — APPROVED** [review]

**Review by ruuddw — APPROVED** [review]

**Review by pabigot — COMMENTED** [review]
Is `samples/tests: Convert use of %ll{u,d} to PRI{u,d}64` intended to be comprehensive?  `git grep '%ll' tests/` finds a few more cases.

**Comment by galak** [comment]
> Is `samples/tests: Convert use of %ll{u,d} to PRI{u,d}64` intended to be comprehensive? `git grep '%ll' tests/` finds a few more cases.

Only the samples/tests that get built on native_posix_64 were touched.

**Comment by galak** [comment]
@carlescufi I suggest we merge this and deal with any Kconfig/relnote fixes a follow up PRs.

**Comment by galak** [comment]
An explicit link the shippable run: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/74656/summary/console

**Review by carlescufi — APPROVED** [review]

**Comment by galak** [comment]
@carlescufi & @nashif opened up PR #26047 to continue the legacy/deprecated naming discussion.

**Pull request merged**